crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs

```diff
@@ -0,0 +1,352 @@
+                range: TextRange::default(),
+                name: Identifier::new(*name, TextRange::default()),
+                default: None,
+            }),
+            TypeParamKind::ParamSpec => TypeParam::ParamSpec(TypeParamParamSpec {
+                range: TextRange::default(),
+                name: Identifier::new(*name, TextRange::default()),
+                default: None,
+            }),
+        }
+    }
+}
+
+struct TypeVarReferenceVisitor<'a> {
+    vars: Vec<TypeVar<'a>>,
+    semantic: &'a SemanticModel<'a>,
+    /// Tracks whether any non-TypeVars have been seen to avoid replacing generic parameters when an
+    /// unknown `TypeVar` is encountered.
+    any_skipped: bool,
+}
+
+/// Recursively collects the names of type variable references present in an expression.
+impl<'a> Visitor<'a> for TypeVarReferenceVisitor<'a> {
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        // special case for typing.AnyStr, which is a commonly-imported type variable in the
+        // standard library with the definition:
+        //
+        // ```python
+        // AnyStr = TypeVar('AnyStr', bytes, str)
+        // ```
+        //
+        // As of 01/2025, this line hasn't been modified in 8 years, so hopefully there won't be
+        // much to keep updated here. See
+        // https://github.com/python/cpython/blob/383af395af828f40d9543ee0a8fdc5cc011d43db/Lib/typing.py#L2806
+        //
+        // to replace AnyStr with an annotation like [AnyStr: (bytes, str)], we also have to make
+        // sure that `bytes` and `str` have their builtin values and have not been shadowed
+        if self.semantic.match_typing_expr(expr, "AnyStr")
+            && self.semantic.has_builtin_binding("bytes")
+            && self.semantic.has_builtin_binding("str")
+        {
+            self.vars.push(TypeVar {
+                name: "AnyStr",
+                restriction: Some(TypeVarRestriction::AnyStr),
+                kind: TypeParamKind::TypeVar,
+                default: None,
+            });
+            return;
+        }
+
+        match expr {
+            Expr::Name(name) if name.ctx.is_load() => {
+                if let Some(var) = expr_name_to_type_var(self.semantic, name) {
+                    self.vars.push(var);
+                } else {
+                    self.any_skipped = true;
+                }
+            }
+            _ => visitor::walk_expr(self, expr),
+        }
+    }
+}
+
+fn expr_name_to_type_var<'a>(
+    semantic: &'a SemanticModel,
+    name: &'a ExprName,
+) -> Option<TypeVar<'a>> {
+    let Some(Stmt::Assign(StmtAssign { value, .. })) = semantic
+        .lookup_symbol(name.id.as_str())
+        .and_then(|binding_id| {
+            semantic
+                .binding(binding_id)
+                .source
+                .map(|node_id| semantic.statement(node_id))
+        })
+    else {
+        return None;
+    };
```

Comment:
I know this is just copied over from where it was before -- but we can simplify this quite a bit with some `?`s:

```suggestion
    let StmtAssign { value, .. } = semantic
        .lookup_symbol(name.id.as_str())
        .and_then(|binding_id| semantic.binding(binding_id).source)
        .map(|node_id| semantic.statement(node_id))?
        .as_assign_stmt()?;
```